### DISCUSSION TITLE
Alternative convolution layers and clustering

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#2892] [Q&A] [answered]
Hi Matt, `SplineConv` actually worked great, and it's pretty straightforward to understand compared to other type of graph convolutions. Thank you for your contribution on that one. 
I'm trying to use other types of graph `conv` layer that outputs `data.x`.

``` python
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = SplineConv(d.num_features, 32, dim=2, kernel_size=5)
        self.conv2 = SplineConv(32, 64, dim=2, kernel_size=5)
        self.fc1 = torch.nn.Linear(64, 128)
        self.fc2 = torch.nn.Linear(128, d.num_classes)

    def forward(self, data):
        data.x = self.conv1(data.x, data.edge_index, data.edge_attr)
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        data = max_pool(cluster, data, transform=transform)

        data.x = self.conv2(data.x, data.edge_index, data.edge_attr)
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        x, batch = max_pool_x(cluster, data.x, data.batch)

        x = global_mean_pool(x, batch)
        x = F.elu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        return F.log_softmax(self.fc2(x), dim=1)
```
In this example, `SplineConv` generates trainable parameters, and outputs new `data.x` using `data.x, data.edge_index, data.edge_attr`. 
Is there any other graph convolution in `PyG` that takes `input arguments` similar to `SplineConv`, and outputs `data.x`?
And is there any other `clustering` algorithm in `PyG` that could be used in this example?

**Comment by rusty1s** [comment, marked as answer]
This is actually true for any GNN implemented in PyG. Every GNN will output new node features `x`, but not all can handle edge features as input. The [cheatsheet](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html) should give you clarification on whether a GNN supports handling of edge features. Let me know if this is helpful to you.

**Reply by hkim716** [reply]
Thanks Matt this is very helpful! I will try to play with other type of convolutions and let you know if I have any question.
